### DISCUSSION TITLE
`ArraySegment<T>` doesn't work well with collection expressions.

### DISCUSSION THREAD

**Discussion by CyrusNajmabadi** [dotnet/csharplang#8164] [Language Ideas]
Forked from an issue raised on dotnet/roslyn here: https://github.com/dotnet/roslyn/issues/73738

The main problem is that one can successfully compiler `ArraySegment<T> a = [];` without warning, while also getting an invalid (`default`) array segment back which explodes on virtually all operations.

The core issue here is exactly what we ran into with `ImmutableArray<T>`.  Namely that it technically fits the collection-expr pattern.  It can be new'ed up, though you get hte equivalent of `new ArraySegment<T>()` which is the same as the `default` value of it.

This discussion is being opened to help track this and discuss potential ways we could solve this.  IMO, the most ideal way to support this would be to relax CollectionBuilderAttribute to allow it to point at a method taking a `T[]` not just a `ReadOnlySpan<T>`.  Seeing that, the compiler would create an array from the collection-expression, and pass that to the creation method.

We want to avoid the ROS approach as that would then force the impl to make a copy of teh span (which may already be wrapping an array the compiler created).

**Comment by RikkiGibson** [comment]
I thought we just had a separate issue where we said we'd like to add support for `Create(T[])` post C#-13. Now I can't find it..

**Reply by jnm2** [reply]
https://github.com/dotnet/csharplang/issues/8144

**Comment by jnm2** [comment]
It was this, or falling back to creating `T[]` when targeting a type with an implicit conversion from `T[]`. (@stephentoub)

However, we do want a "breaking change" where `(ArraySegment<T>)[]` starts to mean something more sensible than what it compiles and runs as today, so adding a fallback alone wouldn't solve the issue.

I also lean toward extending Create support to other types such as `T[]` and `IEnumerable<T>` (for existing dictionary CreateRange methods), regardless of whether we do or don't eventually allow targeting types which are implicitly convertible from `T[]`.

**Comment by huoyaoyuan** [comment]
Even `ImmutableArray<T>` and `List<T>` can benefit from this, or any collection type that uses array for backing storage. The array can guaranteed to be safe to hold (unique unless empty). Then `ImmutableCollectionsMarshal.AsImmutableArray` can be standard creation method, rather than some secret optimization.

**Comment by jnm2** [comment]
I'm championing a feature that would address this issue by allowing collection expressions to use a Create method that takes an array instead of a span. Because this is a concrete solution that addresses multiple separate issues at once, for which there are multiple existing discussions, the proposal discussion is new: https://github.com/dotnet/csharplang/discussions/9924
